### PR TITLE
Merge tests/drivers/adc/adc_api and adc_dma in a single test

### LINKED ISSUE #68690: Merge tests/drivers/adc/adc_api and adc_dma in a single test
**Is your enhancement proposal related to a problem? Please describe.**

I'd like to expand test support for ADC DMA feature for STM32 following this PR: #68685
But the tests/drivers/adc/adc_dma test still configures the ADC channels in the source code instead of in dts overlays (contrary to adc_api that was updated in #56647).
By looking at both tests (and apart from the dts/source code channel config), they seem pretty identical in what they do and test, the difference is that adc_dma enables the DMA (obviously!).

**Describe the solution you'd like**

Since both tests are practically the same, I think it would be a good idea to add a new .conf file and a new test case inside adc_api testcase.yaml that would test the ADC with DMA.
That way, we could remove the duplicate test adc_dma, and reuse the many dts overlays already defined in adc_api by just adding the DMA configuration to them.

**Describe alternatives you've considered**

Another solution could be to port adc_dma test to use the dts overlays, just like adc_api. But the test itself would practically be a duplicate of adc_api, and we would also have to duplicate all overlays.

@anangl I'd be happy to do the change, but I'd like to hear from you before starting on it. Do you think this proposal is acceptable?

### PR DESCRIPTION
**Pull request by gautierg-st**
Tests adc_api and adc_dma being very similar, this PR integrates the DMA test in adc_api test to avoid duplicated code.

Also adds overlays for many STM32 boards.
This PR depends on #69543 for all STM32 tests to pass.

Fixes #68690

### CODE REVIEW

**Comment by gautierg-st** [comment]
I haven't been able to test the change on Kinetis boards.
@hakehuang I think you added the two Kinetis boards on adc_dma test, do you think you could test this PR to see if this still works?
DNM until these tests are passed

**Review by erwango — CHANGES_REQUESTED** [review]
On the last commit, should you update test .yaml then ?

File: tests/drivers/adc/adc_api/overlay-dma-stm32.conf [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#
+# Copyright (c) 2024 STMicroelectronics
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+CONFIG_TEST_USERSPACE=n
+CONFIG_ADC_STM32_DMA=y
+CONFIG_DMA=y
```
[the remark this answers is not in the document]
Comment (gautierg-st) [reply tests/drivers/adc/adc_api/overlay-dma-stm32.conf]: Removed

**Comment by gautierg-st** [comment]
> On the last commit, should you update test .yaml then ?

Correct, I forgot to update the list. Done now!

**Review by erwango — COMMENTED** [review]
Otherwise LGTM

File: tests/drivers/adc/adc_api/overlay-dma-stm32.conf [hunk 2]
```diff
@@ -4,5 +4,5 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
+CONFIG_TEST_USERSPACE=n
 CONFIG_ADC_STM32_DMA=y
-CONFIG_ADC_ASYNC=y
```
[the remark this answers is not in the document]
Comment (gautierg-st) [reply tests/drivers/adc/adc_api/overlay-dma-stm32.conf:8]: `CONFIG_ADC_ASYNC=y` is already included in prj.conf, so the async test is also done with DMA

**Review by erwango — DISMISSED** [review]

**Comment by hakehuang** [comment]
kinetis testing fails
```
bb*** Booting Zephyr OS build v3.6.0-220-g50dbae9b2c32 ***
Running TESTSUITE adc_basic
===================================================================
START - test_adc_asynchronous_call
Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/adc/adc_api/src/test_adc.c:188: test_task_asynchronous_call: (ret not equal to 0)
k_poll failed with error -11
FAIL - test_adc_asynchronous_call in 1.015 seconds
===================================================================
START - test_adc_invalid_request

```
it hangs here

**Comment by gautierg-st** [comment]
> it hangs here

Thanks for your test. I didn't notice that Kinetis SoCs used a counter. It's added now, along with the sample interval.
I've checked the compilation but could you test again on frdm_k64f and frdm_k82f?

**Comment by hakehuang** [comment]
> Thanks for your test. I didn't notice that Kinetis SoCs used a counter. It's added now, along with the sample interval.
> I've checked the compilation but could you test again on frdm_k64f and frdm_k82f?

the drivers.adc.dma_nxp_kinetis case is ok, but I meet problem with the drivers.adc, it fails on both boards.

**Comment by gautierg-st** [comment]
> the drivers.adc.dma_nxp_kinetis case is ok, but I meet problem with the drivers.adc, it fails on both boards.

I finally got my hand on a frdm_k64f board and was able to test and fix the issue. I've tested both with and without DMA and it works. Feel free to confirm if you want.

File: tests/drivers/adc/adc_api/src/test_adc.c [hunk 3]
```diff
@@ -18,8 +20,18 @@
...
 
+#if CONFIG_NOCACHE_MEMORY
+#define __NOCACHE	__attribute__((__section__(".nocache")))
+#else /* CONFIG_NOCACHE_MEMORY */
+#define __NOCACHE
+#endif /* CONFIG_NOCACHE_MEMORY */
+
 #define BUFFER_SIZE  6
+#ifdef CONFIG_TEST_USERSPACE
 static ZTEST_BMEM int16_t m_sample_buffer[BUFFER_SIZE];
```
Comment (erwango) [inline tests/drivers/adc/adc_api/src/test_adc.c:31]: No need to define this memory as non cacheable ?

**Review by erwango — DISMISSED** [review]

File: tests/drivers/adc/adc_api/src/test_adc.c (hunk 3 above)
Comment (gautierg-st) [reply tests/drivers/adc/adc_api/src/test_adc.c:31]: No. The case with CONFIG_TEST_USERSPACE is for the basic adc_api test, without DMA. It has worked without like this for a long time, and setting nocache attribute with USERSPACE enabled causes MPU fault anyway.
For the test with DMA, we have to disable USERSPACE to define the nocache attribute.

**Comment by hakehuang** [comment]
> I finally got my hand on a frdm_k64f board and was able to test and fix the issue. I've tested both with and without DMA and it works. Feel free to confirm if you want.

now it works for me.

**Comment by erwango** [comment]
@gautierg-st Please rebase

**Review by erwango — APPROVED** [review]

**Review by decsny — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Pull request merged**